### DISCUSSION TITLE
areaY stacked problem with pos/neg values

### DISCUSSION THREAD

**Discussion by Hvass-Labs** [observablehq/plot#1963] [Ideas]
When data is plotted using `areaY` that contains both positive and negative values, it creates very confusing zig-zag lines due to the linear interpolation between pos/neg values. It took me a moment to realize what was happening when I first saw this, and I don't think the average user of my web-site will understand this. So I hope you can remove this artifact. Thanks!

Example: https://observablehq.com/d/a1c4060fcfcc75b4

PS: Your HQ Notebook system is amazing! I just wish it supported Python as well.

![data1](https://github.com/observablehq/plot/assets/13588114/b0ed097f-4d0a-4715-b9c6-bbac0c68f000)
![data2](https://github.com/observablehq/plot/assets/13588114/87d69f58-c53b-45f8-a1bc-c49b393eeab0)

**Comment by Fil** [comment]
Areas are continuous, which explains the problem; so maybe try instead:
- a different mark (for example `Plot.rectY(data1, {x: "x", y: "y", stroke: "z", fill: "z", interval: "3 months"})`)
- keep the areaY mark, but separate positive and negative series, using an interval to make sure there are 0 values when necessary: 
```
  Plot.areaY(data1, {
      x: "x",
      y: "y",
      fill: "z",
      z: (d) => `${Math.sign(d.y)} ${d.z}`,
      interval: "3 months"
    }),
```

- keep the areaY mark, separate positive and negative series by duplicating the dataset:
```js
    Plot.areaY(
      data1.flatMap((d) => [
        { ...d, y: Math.max(0.00001, d.y), series: `positive ${d.z}` },
        { ...d, y: Math.min(-0.00001, d.y), series: `negative ${d.z}` }
      ]),
      {
        x: "x",
        y: "y",
        fill: "z",
        z: "series"
      }
    ),
```

**Comment by mbostock** [comment]
To add to @Fil’s answer, here is complete working code. I would probably apply the **interval** option to the _x_ scale rather than the rect mark so that you get an ordinal axis, but it’s really a stylistic preference as to whether you want to draw the bars between the ticks or centered on the ticks (and hence whether you want to represent time as a continuous interval or discrete periods).

![untitled - 2023-12-27T104255 298](https://github.com/observablehq/plot/assets/230541/f159cbb7-3af2-404b-ace3-2362b87957b8)

```js
Plot.plot({
  x: {interval: "quarter", label: null},
  y: {grid: true},
  color: {legend: true},
  marks: [
    Plot.rectY(data1, {x: "x", y: "y", fill: "z", tip: true}),
    Plot.ruleY([0])
  ]
})
```

Another simple solution to this problem is to avoid interpolation by using the _step_ **curve** instead of the default _linear_.

![untitled - 2023-12-27T104656 834](https://github.com/observablehq/plot/assets/230541/98341734-9e35-4882-adaa-8cc9950ff229)

```js
Plot.plot({
  y: {grid: true},
  color: {legend: true},
  marks: [
    Plot.areaY(data1, {x: "x", y: "y", curve: "step", fill: "z"}),
    Plot.ruleY([0])
  ]
})
```

**Comment by Hvass-Labs** [comment]
Thanks for your generous replies!

These are monetary amounts that sometimes switch between positive and negative, so the zig-zag lines would be rather meaningless and impossible to interpret. It is only because I understand that they are artifacts of interpolation that I can try and ignore them. But the ordinary user won't understand that and would be quite confused about the zig-zags.

But I prefer using lines as opposed to bars / steps because I think the lines look very pretty here :-)

@Fil's last suggestion is exactly what I expect to see. I have updated the Notebook to show this desired result: https://observablehq.com/d/a1c4060fcfcc75b4

Ideally this would be done automatically "behind-the-scenes" so I don't have to worry about it. It should also work with brushing / selection of data.

Would it be possible to add this to the `areaY` function?

Thanks!

![solution1](https://github.com/observablehq/plot/assets/13588114/fb32f155-ecd6-4d98-acd3-52470f31307c)

**Comment by mbostock** [comment]
If you like @Fil’s solution, I would recommend a different implementation that is simpler and avoids the need for near-zero values (such as 0.0001). Essentially, this duplicates the stacked area so that there is a positive stacked area (growing upwards) and a negative stacked area (growing downwards), rather than alternating the two.

![untitled - 2023-12-29T120738 064](https://github.com/observablehq/plot/assets/230541/8e86386a-4317-4de0-9576-8ce3bf4acc33)

```js
Plot.plot({
  y: {grid: true},
  color: {legend: true},
  marks: [
    Plot.areaY(data1, {x: "x", y: (d) => Math.max(0, d.y), fill: "z"}),
    Plot.areaY(data1, {x: "x", y: (d) => Math.min(-0, d.y), fill: "z"})
  ]
})
```

Personally, I prefer the rect or _step_ curve approach to avoid interpolation in this case.

**Reply by Hvass-Labs** [reply]
Thanks again for your help!

I tried with data that has ten different strings for "z". Using your solution I still got zig-zag lines in the negative y-region, so I had to use this code-line instead:

    Plot.areaY(data1, {x: "x", y: (d) => Math.min(-1e-35, d.y), fill: "z"})    

I tried using the value -0.0 instead of -1e-35 but that still gave zig-zag lines.

The plot looks nice and I can certainly live with the tiny hack. But I wonder if there is a small bug in the plotting code that requires this? It's a bit of a hassle for me to dump this data into a CSV-file, but if you need it to help diagnose the problem, then I can do it to help you out.

Here's the problem when using a normal `areaY` mark:

![many-zig-zag](https://github.com/observablehq/plot/assets/13588114/5cd08681-2400-4600-9d87-d3c9885d29e2)

Here's the problem for the negative region when using your suggestion with two `areaY` marks and `Math.min/max` with zero:

![many-zig-zag-neg](https://github.com/observablehq/plot/assets/13588114/ff4d04b5-58df-4ab6-8a4f-4739191bdcff)

Here's my tiny hack to your solution using the constant `-1e-35` instead of zero:

![many-solution](https://github.com/observablehq/plot/assets/13588114/418096a5-a39b-4603-be5c-30879f5b714d)

Here's the solution using `rectY` which I think looks a bit crowded with this data:

![many-rectY](https://github.com/observablehq/plot/assets/13588114/a05c1305-983e-44a5-88da-08da714bc140)

Here's the solution using `areaY` with `curve: "step"` which I think looks a bit pixelated:

![many-areaY-curve-step](https://github.com/observablehq/plot/assets/13588114/d37e4271-37e8-45b3-aed0-576e61d94322)

Personally I prefer the `areaY` with lines once the zig-zags are removed.

Happy New Year! :-)

**Reply by mbostock** [reply]
Nice, yes, that’s a bug. We can make `Math.min(-0, d.y)` work as intended here. https://github.com/observablehq/plot/pull/1967

**Reply by Hvass-Labs** [reply]
I'm sorry I first saw your updates now. You requested the data in the PR. Please see the updated Notebook: https://observablehq.com/d/a1c4060fcfcc75b4

The hack works better on my local computer where the result is as shown above. I use a beta-build based on v0.6.11 that has the brushing feature.

Please also note that this hack messes with the stroke-color.

I wonder if a better solution might be to handle the switching between pos/neg stacked regions internally in the `areaY` code, instead of hacking it this way?

**Reply by Fil** [reply]
Thanks for sharing! It's very interesting to see this dataset, because it has a second peculiarity, which is that two series are labelled B (one with positive numbers, the other with negative numbers). If we sort by x, this results in weird lines such as this one:
![](https://github.com/observablehq/plot/assets/7001/04b712ed-2290-4eaa-b0b7-dfc228a76831)

This can be solved by grouping with a sum reducer (I think? We don't know what the data mean nor why there are two series, so it's still a wild guess), which works well with #1968:
![Capture d’écran 2024-01-04 à 22 27 17](https://github.com/observablehq/plot/assets/7001/d67cb191-eba2-4cbd-9d9b-86ad48a59860)

**Reply by Hvass-Labs** [reply]
> It's very interesting to see this dataset, because it has a second peculiarity, which is that two series are labelled B (one with positive numbers, the other with negative numbers).

Oops! I'm terribly sorry I made a mistake! It is such a hassle to extract this data that I accidentally used the name "B" twice. That also explains why it looked worse than the other plots I made above. I've fixed it in the [Notebook](https://observablehq.com/d/a1c4060fcfcc75b4) and checked it should be correct now. I'll be more careful the next time. But at least you found it interesting so it wasn't a complete waste of your time :-)

**Reply by Hvass-Labs** [reply]
By the way, these are all monetary amounts that change over time. Most of them can change between pos/neg.

**Comment by Fil** [comment]
> a better solution might be to handle the switching between pos/neg stacked regions internally in the areaY code

Yes, maybe in the stackY transform, where values are grouped by z, they could be further duplicated with alternating zeroes… might be a lot of code overindexing on this particular example, though 🤔

**Reply by Hvass-Labs** [reply]
Note that the hack with two `areaY` marks also messes with the stroke-color. The hack is still a satisfactory solution for me and I will use it for now. But from the user's perspective it appears to be an artifact / bug in the plotting code, so eventually it would be nice with a proper solution. For monetary values that can change between positive and negative over time, this would be a common problem.

My naive idea for a solution would be to pre-process the data inside `areaY` or the `stackY` transform, and separate it into pos/neg values. Perhaps it is necessary to also insert "phantom" points for zero and negative-zero when the data switches between pos/neg, but I wonder if that would mess with the brushing / selection feature. But I've never written plotting routines so I don't know exactly how all this works :-)
